flocker/control/_persistence.py

```diff
@@ -21,16 +21,94 @@
 
+# The latest configuration version. Configuration versions are
+# always integers.
+_CONFIG_VERSION = 2
 
-class _ConfigurationEncoder(JSONEncoder):
+# Map of serializable class names to classes
+_CONFIG_CLASS_MAP = {cls.__name__: cls for cls in SERIALIZABLE_CLASSES}
+
+
+class ConfigurationMigrationError(Exception):
+    """
+    Error raised when a configuration migration is unable to take place
+    or complete successfully.
+    """
+
+
+def migrate_configuration(source_version, target_version,
+                          config, migration_class):
+    """
+    Migrate a persisted configuration from one version to another
+    in sequential upgrades, e.g. a source version of 1 and target
+    version of 3 will perform two upgrades, from version 1 to 2,
+    followed by 2 to 3.
+
+    Calls the correct ``ConfigurationMigration`` class methods for
+    the suppled source and target versions.
+
+    :param int source_version: The version to migrate from.
+    :param int target_version: The version to migrate to.
+    :param bytes config: The JSON-encoded source configuration.
+    :param class migration_class: The class containing the methods
+        that will be used for migration.
+
+    :return bytes: The updated JSON configuration after migration.
+    """
+    upgraded_config = config
+    current_version = source_version
+    for upgrade_version in range(source_version + 1, target_version + 1):
+        with _LOG_UPGRADE(configuration=upgraded_config,
+                          source_version=current_version,
+                          target_version=upgrade_version):
+            migration_method = (
+                u"configuration_v%d_v%d"
+                % (current_version, upgrade_version)
+            )
+            try:
+                migration = getattr(migration_class, migration_method)
+            except AttributeError:
+                message = (
+                    u"Unable to find a migration path for a version " +
+                    unicode(source_version) + u" to version " +
+                    unicode(target_version) + u" configuration. " +
+                    u"No migration method exists for v" +
+                    unicode(current_version) + u" to v" +
+                    unicode(upgrade_version) + u"."
+                )
+                raise ConfigurationMigrationError(message)
+            upgraded_config = migration(upgraded_config)
+            current_version = current_version + 1
+    return upgraded_config
+
+
+class ConfigurationMigration(object):
+    """
+    Migrate a JSON configuration from one version to another.
+    """
+    @classmethod
+    def configuration_v1_v2(cls, config):
+        """
+        Migrate a v1 JSON configuration to v2.
+
+        :param bytes config: The v1 JSON data.
+        :return bytes: The v2 JSON data.
+        """
+        v2_config = Configuration(
+            version=2, deployment=wire_decode(config))
+        return wire_encode(v2_config)
```

Comment:
Using `wire_decode` here will eventually lead to problems.  This use only works as long as the model objects in the current version of the code can be deserialized from the older version of the configuration.  This works for this particular upgrade for the time being but it's more or less guaranteed to break at some point.  Ideally, old upgrade code would be as little of a maintenance burden as possible.  So having old upgraders break when someone is developing a new feature isn't ideal.

We can be fairly confident that the `json` module isn't going to undergo any drastic changes that render it incapable of deserializing the configuration to some mid-level Python objects, though.  Those mid-level Python objects are probably the right level at which upgrades can operate.  That means this upgrade method should look something like:

```py
v1_config = json.loads(config)
v2_config = {"$__class__$": ..., "deployment": v1_config}
return json.dumps(v2_config)
```

The details of the most robust/reliable way to construct the `v2_config` wrapper/skeleton might vary a bit - but in this case it's so simple there's not really much choice, I think.